# More `numpy`

A more complete discussion of the basic use of `numpy` is given [here](http://www.scipy-lectures.org/intro/numpy/operations.html). These are all things that, based on my experience, you will end up using all the time. This can still only cover the most basic operations -- there is a huge amount of stuff in `numpy`.

- Differences between copies of arrays and references to arrays
- Sorting arrays
- Concatenating, stacking and splitting arrays
- Creating arrays with `repeat`, `reshape`
- Record arrays (recarrays)
- Logical masks and crazy floating point values (`nan` and `inf`)
- `numpy.where` with 2d arrays
- Finding unique values
- Testing if values from one array are in another array, using in1d
- Finding locations of specific values in arrays

## Copies vs. references

We covered this for basic Python data types last week, and saw the `copy` module. Numpy arrays have their own in-built `copy` function which does the same thing, and is the best way to copy them.

In [ ]:
x = np.zeros(3,dtype=np.float32)
y = x
z = x.copy()
x[2] = 100.0 # set an element of x

print('x = {}'.format(x))
print('y = {}'.format(y))
print('z = {}'.format(z))
print('y is x? {}'.format(y is x))
print('z is x? {}'.format(z is x))

The point is that `y=x` makes `y` a reference to `x`, rather than a copy, so updating `y` also updates `x`. In practice it can get quite hard to guess whether any of the hundreds of functions in `numpy` will return a copy or a reference to the array you give it as input (in some cases this even depends on the input!).

Bottom line: when you really want a copy, explicitly make a copy.

## Sorting arrays

There are two simple ways to sort an array `x`: either `np.sort(x)`, which returns a copy of `x` in sorted order, or `x.sort()`, which sorts x 'in place' -- it actually moves the values in `x` itself into sorted order.

This choice between `np.something(x)` (copy) and `x.something()` (in place change to `x`) is similar for a lot of `numpy` functions.

In [ ]:
np.random.seed(42)
x = np.random.randint(0,100,size=10) # Some random integers
print('Original x: {}'.format(x))
print('np.sort(x): {}'.format(np.sort(x))) # np.sort returns a copy
print('         x: {}'.format(x))
print('  x.sort(): {}'.format(x.sort()))   # x.sort() returns None...
print('         x: {}'.format(x))          # but now x is sorted!

In [ ]:
np.random.seed(42)
x = np.random.randint(0,100,size=10) # Some random integers
a = np.argsort(x)
print(a)          # The *indices* rather than the *values* in sorted order
print(x[a])       # Use the indices to sort the array
print(x[a[::-1]]) # Sort backwards by reversing the indices

This is useful because you can now use the indices `a` to put any other array in the same order, not just `x`.

## Concatenating, stacking and splitting

These are all ways to combine multiple arrays into a single array. `vstack` and `hstack` add rows and columns respectively, increasing the dimension of the array (i.e. making two 1d arrays into one 2d array). `vstack` means 'vertical stack' (add more rows) and `hstack` means horizontal stack (add more columns) -- experiments are the easiest way to understand what the difference is. 

`concatenate` joins arrays end-to-end along a particular dimension (by default, the first one). For example, it's the standard way to combine a list of 1d arrays into one big 1d array.

In [ ]:
x = np.arange(10,25)
y = np.arange(30,45)

print('x = {}'.format(x))
print('y = {}'.format(y))
print('concatenated: {}'.format(np.concatenate((x,y))))

v = np.vstack((x,y))
print('shape after vstack: {}'.format(u.shape))

h = np.hstack((x,y))
print('shape after hstack: {}'.format(v.shape))

# In this case, stacking horizontally is the same as concatenating 
print(np.all(h == np.concatenate([x,y])))

There are also very similar functions to split up arrays into smaller arrays. Following on from the previous cell:

In [ ]:
a,b = np.vsplit(v,2)
np.all(a==x) and np.all(b==y) # We should be back to where we started!

## Creating arrays, repeat, reshape

There are [loads of different functions](https://docs.scipy.org/doc/numpy/reference/routines.array-creation.html) for creating arrays in `numpy`. These include `zeros` to make an empty array and `repeat` to make an array where every element has a particular value. We saw these last week.

Let's examine how long it takes to create a big array of -1's, using IPython's `%timeit` magic function.

In [ ]:
%timeit np.zeros(1000) - 1

In [ ]:
%timeit np.repeat(-1,1000)

So in this case, the first way is about 3x faster. This is helpful to know if you're making very big arrays. If you can think of two ways to do something, then either is fine, unless one of them **really** slows down your program. 

Sometimes we want to make an array with a repeating pattern, like `[1,0,-1,1,0,-1,1,0,0,-1]`. Let's try to do that with repeat:

In [ ]:
np.repeat((1,0,-1),10)

Clearly that's not the right way to do it. The correct function in this case is `tile`:

In [ ]:
np.tile((1,0,-1),10)

The `reshape` function can be used if we want a 2d version

In [ ]:
np.reshape(np.tile((1,0,-1),10), (10,3))

## Record arrays (recarrays)

You can define your own numpy data types that are combinations of the basic types, with each element of the new composite type associated with a name. This is useful for storing **structured data**, and the most general way to do that with `numpy` (although in many cases you might find the Table class from `astropy` easier to work with).

The `line_data` we defined in the first cell is a structured array:

In [ ]:
line_data[0:3] # Show the first three elements 

In [ ]:
# You can see this has a complicated dtype, made up of 6 elements
line_data.dtype

In [ ]:
# line_data is **not** a 20x6 array, it's a 20x1 array of elements of this dtype.
line_data.shape

In [ ]:
# If you want to know the total number of bytes in each element, you can:
line_data.dtype.itemsize

*The important thing* is that individual fields can be accessed by name:

In [ ]:
line_data['ID']

To make a record array like this from scratch, use the function `numpy.dtype`, passing a ***list of tuples*** of the form (`name_of_field`, `data_type_of_field`). The data type can either be an existing `dtype`, like `numpy.int32`, or a string. `numpy` will convert the string into a dtype, if it can. The '<' and '>' indicated 'little endian' or 'big endian' representations. For example


In [ ]:
my_dtype = np.dtype([('ID', '|S16'), ('ra', '<f8'), ('dec', '<f8')])
np.zeros(4,dtype=my_dtype) # Make an empty array with four of these elements

The `|S16` is the way to specify a string of length 16 characters as a datatype. 

## Logical masks and crazy floating point values (`nan` and `inf`)

We can make an array of booleans that indicate whether each element in the array meets some criterion we define. These are often called 'masks'.

In [ ]:
mask = (line_data['x'] > 200.0) & (line_data['y'] > 500.0)
mask

We can 'filter' the array by using the mask as an index.

In [ ]:
line_data[mask]

As we saw last week, we can use `numpy.where` to find the actual indices of the elements for which the mask is `True`:

In [ ]:
w = np.where(mask)[0]
print(w)
line_data[w]

Very often you'll end up doing something like taking the `log10` of an array with some negative numbers or dividing an array by another array with zeros in it. In those cases, the special values `nan` (not a number), `inf` (infinity) and `-inf` (negative infinity) will show up in the results, mixed in with the elements where whatever operation you tried *was* mathematically valid. For example:

In [ ]:
crazy_values = np.log10(np.tile((1,0,-1),10)) # This will cause some warnings to appear!
print(crazy_values[0:3])

`numpy` has some routines to make masks of the elements that contain these special values.

In [ ]:
np.where(np.isinf(crazy_values))

In [ ]:
np.where(np.isnan(crazy_values))

The `isfinite` routine returns True for the elements in its argument that are *not* any of these special values.

In [ ]:
np.where(np.isfinite(crazy_values))

___
Further reading:

For advanced/frequent work with masked arrays, there is also the `numpy.ma` module, which defines a new 'masked array' type. This gets quite complicated to work with and I don't use it very often. It will be more useful if you are doing complicated linear algebra.

##  `numpy.where` with 2d arrays

Last week we saw that (for example) `np.where(x>something)` returns a tuple of length 1 if `x` is a 1d array, so to get the actual elements of x that are `>something` we need to get the first (and only) element of the result: `np.where(x>something)[0]`.

What if `x` is 2d?

In [ ]:
x = np.arange(9).reshape((3,3))
print(x.shape)
w = np.where(x > 4)
print(w) # note this is a tuple of length 2
print(x[w])

## Unique values

To find the unique values in an array, use `numpy.unique`.

In [14]:
x = np.random.randint(0,10,20) # 20 random integers between 0 and 10
print(x)
print(np.unique(x))

[0 3 7 6 8 2 6 8 3 4 2 3 1 2 4 3 7 5 1 6]
[0 1 2 3 4 5 6 7 8]


Note that the results are sorted. The `numpy.unique` function has some useful optional arguments:

In [15]:
u, idx, inv, counts = np.unique(x,return_index=True,return_inverse=True,return_counts=True)

You can check what these do yourself. One useful application is to make a list of lengths and offsets of values in a sorted array. Here's a trivial example.

In [16]:
x = np.random.randint(0,10,200) # 20 random integers between 0 and 10
a = np.argsort(x)
values, offset, length = np.unique(x[a],return_index=True,return_counts=True)
for u,o,l in zip(values,offset,length):
    print('{} | {}'.format(u, x[a][o:o+l] ))

0 | [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
1 | [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
2 | [2 2 2 2 2 2 2 2 2 2 2 2 2 2]
3 | [3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3]
4 | [4 4 4 4 4 4 4 4 4 4 4 4 4 4 4]
5 | [5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5]
6 | [6 6 6 6 6 6 6 6 6 6 6 6 6 6 6 6 6 6 6]
7 | [7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7]
8 | [8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8]
9 | [9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9]


Now imagine that `x` is a label that groups some other properties: for example, the 'galaxy ID' of particles in a cosmological simulation or the 'source ID' of multiple exposures of a set of targets observed with a spectrograph. The lengths and offset arrays can be used to find the properties of the subsets defined by `x`:

In [17]:
luminosity = np.random.random(200) # Some property associated with each x from the last cell

print('  | Total Luminosity')
for u,o,l in zip(values,offset,length):
    print('{} | {}'.format(u, np.sum(luminosity[a][o:o+l])))


  | Total Luminosity
0 | 8.77927438261
1 | 8.56983425818
2 | 6.73454780353
3 | 10.4039925833
4 | 7.56799804489
5 | 13.7637569337
6 | 10.4030902649
7 | 11.2809670418
8 | 12.7977817309
9 | 10.6023711337


There are even more general ways to do something similar (the `pandas` package has many such functions) and more efficient techniques with `numpy`, to get rid of that `for` loop (too advanced for this tutorial, but if you want more details have a look at the documentation for `numpy.ufunc.reduceat` and google for some examples, e.g. of `add.reduceat`). The above example works well in most simple cases. Remember the array has to be sorted!

## Testing if values from one array are in another array, using in1d

`numpy.in1d` provides a quick way to check if each element of an array is in another array, but it doesn't tell you *where* those elements are in the other array.

In [18]:
odd_numbers = np.arange(1,100,2)
np.in1d([1,2,3,4,5],odd_numbers)

array([ True, False,  True, False,  True], dtype=bool)

Efficiently finding the index of each matching element in the other array is the point of the next example.

## Finding locations of specific values in arrays

Matching specific values in very long lists of arbitrary integers is a common task; for example, in working with the particle data from cosmological simulations, you might might have a list of particle IDs in one snapshot of the simulation, and want to find the same particles in another snapshot.

Currently, this operation isn't covered by an existing routine in `numpy` or `scipy` (as far as I know!). The nearest thing is in the `pandas` statistics package, which we'll look at next week. However, it's not *too* hard to write such a routine yourself. This is the basis of one of the challenges.

First, we'll make an array of 1000 unique integer values, `BIG_ARRAY`. Since we want the values to be unique but not sorted for this example, this takes a couple of steps to set up:

In [19]:
np.random.seed(42)
x         = np.random.randint(0,500,200) # 200 random integers between 0 and 500
BIG_ARRAY = np.unique(x)

# Note this trick for getting a random ordering by sorting random numbers
random_order = np.argsort(np.random.random(len(BIG_ARRAY)))
BIG_ARRAY    = BIG_ARRAY[random_order]

BIG_ARRAY    = BIG_ARRAY + 2000 
print(BIG_ARRAY)

[2054 2330 2149 2409 2100 2474 2454 2171 2217 2034 2049 2498 2128 2201 2062
 2091 2254 2328 2339 2269 2458 2008 2475 2241 2041 2095 2213 2252 2392 2178
 2088 2105 2359 2102 2484 2259 2134 2459 2087 2052 2270 2418 2313 2489 2032
 2061 2315 2455 2348 2491 2048 2212 2435 2151 2216 2385 2058 2050 2483 2343
 2174 2306 2461 2463 2226 2190 2099 2282 2138 2043 2379 2162 2326 2040 2417
 2188 2071 2460 2319 2372 2337 2397 2205 2351 2257 2230 2161 2064 2004 2243
 2350 2443 2166 2098 2264 2401 2445 2303 2363 2169 2014 2001 2027 2156 2358
 2387 2053 2389 2300 2476 2236 2160 2080 2020 2251 2206 2106 2273 2214 2121
 2345 2240 2344 2427 2406 2295 2200 2187 2215 2191 2235 2466 2276 2492 2292
 2279 2207 2391 2021 2267 2047 2419 2288 2378 2051 2471 2293 2308 2221 2013
 2430 2263 2309 2135 2366 2448 2256 2189 2327 2413 2260 2130]


I generated 200 random integers between 0 and 500, find only the unique values, then randomize their order. To make it clear that the numbers can be anything for this example, I added 2000 to each of them. 

Now let's say we wanted to find where the value 2059 was in this array. Obviously we could use `where`:

In [20]:
np.where(BIG_ARRAY == 2201)[0]

array([13])

and if we wanted to find more values, we might use a loop:

In [21]:
# Reset the seed, so we get consistent results each time we try this
np.random.seed(232)
small_array = np.random.randint(2100,2300,10) # some values to look for: not all of them will be found!

# Loop
matches = list()                                    
for s in small_array:
    w = np.where(BIG_ARRAY == s)[0]
    if len(w) == 0:
        # If we didn't find anything, store -1
        matches.append(-1)
    else:
        matches.append(w[0])
    
matches = np.array(matches)
print(matches)

[ -1  31 161  84  -1  -1  92 146  -1  -1]


However, if either `small_array` or `BIG_ARRAY` are large, this is a **very slow** way to do it, because `np.where` has to compare each value from `small_array` against *every* value in BIG_ARRAY. In most of these cases, we didn't even find a match.

**Solution:** When we're dealing with large arrays, it's usually more efficient to sort `BIG_ARRAY` and use a [**bisection search**](https://en.wikipedia.org/wiki/Binary_search_algorithm). In `numpy` we can do this with the `searchsorted` function.

In [22]:
# First sort BIG_ARRAY
sorting_idx = np.argsort(BIG_ARRAY)

In [23]:
# Now use searchsorted (have a look at the docstring for this routine as well)

# What is the *smallest* index in the sorted BIG_ARRAY where I can insert the value '2101'
# without changing the order of BIG_ARRAY?
print(BIG_ARRAY[sorting_idx].searchsorted(2201,side='left'))

# What is the *largest* index in the sorted BIG_ARRAY where I can insert the value '2101'
# without changing the order of BIG_ARRAY?
print(BIG_ARRAY[sorting_idx].searchsorted(2201,side='right'))

60
61


In [24]:
# Just to demonstrate that's the case, show the indices around 56-57
BIG_ARRAY[sorting_idx][[59,60,61,62]]

array([2200, 2201, 2205, 2206])

Notice that doesn't directly tell us which of the indices (56 or 57) is *equal* to  2189, or even that **either** of them are. However, look what happens when we try to find a value that isn't in `BIG_ARRAY`:

In [25]:
print(BIG_ARRAY[sorting_idx].searchsorted(2202,side='left'))
print(BIG_ARRAY[sorting_idx].searchsorted(2201,side='right'))

61
61


This should be enough information to get started with the 'matching values in huge arrays of integers' challenge.

## End of notebook